### PR TITLE
Fix track2 ga naming and client merge issues

### PR DESCRIPTION
**Pull request by qiaozha**

### CODE REVIEW

File: test/integration/generated/datalakestorage/src/models/index.ts [hunk 1]
```diff
@@ -180,7 +180,7 @@ export interface PathUpdateHeaders {
   /** User-defined properties associated with the file or directory, in the format of a comma-separated list of name and value pairs "n1=v1, n2=v2, ...", where each value is a base64 encoded string. Note that the string may only contain ASCII characters in the ISO-8859-1 character set. */
   properties?: string;
   /** When performing setAccessControlRecursive on a directory, the number of paths that are processed with each invocation is limited.  If the number of paths to be processed exceeds this limit, a continuation token is returned in this response header.  When a continuation token is returned in the response, it must be specified in a subsequent invocation of the setAccessControlRecursive operation to continue the setAccessControlRecursive operation on the directory. */
-  xMsContinuation?: string;
```
Comment (qiaozha) [inline test/integration/generated/datalakestorage/src/models/index.ts:183]: This is because Style.camel("XMsContinuation", true, {}, 5} gives us XMsContinuation

File: test/integration/generated/datalakestorage/src/models/index.ts [hunk 2]
```diff
@@ -82,7 +82,7 @@ export interface BlobItemInternal {
   deleted: boolean;
   snapshot: string;
   versionId?: string;
-  isCurrentVersion?: boolean;
```
Comment (qiaozha) [inline test/integration/generated/datalakestorage/src/models/index.ts:85]: Not sure why we have this change

File: test/integration/generated/datalakestorage/src/models/index.ts [hunk 3]
```diff
@@ -301,7 +301,7 @@ export interface PathGetPropertiesHeaders {
   /** The POSIX access permissions for the file owner, the file owning group, and others. Included in the response if Hierarchical Namespace is enabled for the account. */
   permissions?: string;
   /** The POSIX access control list for the file or directory.  Included in the response only if the action is "getAccessControl" and Hierarchical Namespace is enabled for the account. */
-  acl?: string;
+  ACL?: string;
```
Comment (qiaozha) [inline test/integration/generated/datalakestorage/src/models/index.ts:304]: probably should use override for this case?

File: test/integration/generated/lro/src/index.ts [hunk 4]
```diff
@@ -7,6 +7,5 @@
  */
 
 export * from "./models";
-export { LROClient } from "./lROClient";
-export { LROClientContext } from "./lROClientContext";
+export { LROClient } from "./LROClient";
```
Comment (qiaozha) [inline test/integration/generated/lro/src/index.ts:10]: Does the file name change look ok ? This should also because of Style.camel("LRO") = "LRO"

File: test/integration/generated/lro/src/models/mappers.ts [hunk 5]
```diff
@@ -198,10 +198,10 @@ export const SubProduct: coreClient.CompositeMapper = {
   }
 };
 
-export const LROsPatch200SucceededIgnoreHeadersHeaders: coreClient.CompositeMapper = {
+export const LROsPatch200SucceededIgnoreHeaders: coreClient.CompositeMapper = {
```
Comment (qiaozha) [inline test/integration/generated/lro/src/models/mappers.ts:201]: Not sure why we change it from HeadersHeaders to Headers ?

File: test/integration/generated/lro/src/operations/index.ts [hunk 6]
```diff
@@ -6,7 +6,7 @@
  * Changes may cause incorrect behavior and will be lost if the code is regenerated.
  */
 
-export * from "./lROs";
-export * from "./lRORetrys";
+export * from "./LROs";
+export * from "./LRORetrys";
 export * from "./lrosaDs";
-export * from "./lROsCustomHeader";
+export * from "./LROsCustomHeader";
```
Comment (qiaozha) [inline test/integration/generated/lro/src/operations/index.ts:12]: The same as the previous one.

File: test/integration/generated/nameChecker/src/models/parameters.ts [hunk 7]
```diff
@@ -340,8 +340,8 @@ export const selectedFields: OperationQueryParameter = {
   collectionFormat: "CSV"
 };
 
-export const search$DONotNormalize$Text: OperationQueryParameter = {
-  parameterPath: "search$DONotNormalize$Text",
+export const searchDONOTNormalizeText: OperationQueryParameter = {
+  parameterPath: "searchDONOTNormalizeText",
```
Comment (qiaozha) [inline test/integration/generated/nameChecker/src/models/parameters.ts:344]: Maybe this is because I set the Style.camel(str, true, {}, 5) but if I don't set it as 5, I get a mismatch in bodyComplex of CMYKColor in models/index.ts.

File: test/smoke/generated/network-resource-manager/src/models/index.ts [hunk 8]
```diff
@@ -775,7 +775,7 @@ export interface BastionShareableLinkListRequest {
 /** Bastion Shareable Link. */
 export interface BastionShareableLink {
   /** Reference of the virtual machine resource. */
-  vm: Vm;
+  vm: VM;
```
Comment (qiaozha) [inline test/smoke/generated/network-resource-manager/src/models/index.ts:778]: Guess this is looking correct?

File: test/integration/generated/datalakestorage/src/models/index.ts (hunk 2 above)
Comment (qiaozha) [reply test/integration/generated/datalakestorage/src/models/index.ts:85]: look like this is also because Style.camel("IsCurrentVersion", true, {}, 3) gives us IsCurrentVersion. Really weird

**Comment by ramya-rao-a** [comment]
@qiaozha The PR description is empty. Can you please update it with details of 
- links to issues you are resolving with this PR
- your approach to resolving those issues
- anything that the PR reviewer needs to know upfront before reviewing the PR

Both @sarangan12 and @joheredi are out this week. So, @deyaaeldeen will be reviewing this PR. Therefore, any context you can provide will be appreciated

**Comment by ramya-rao-a** [comment]
@deyaaeldeen I believe the issues related to this PR are #1131 and #1134

**Comment by qiaozha** [comment]
@deyaaeldeen This PR is trying to resolve issue #1131 #430 and #1134.
For issue #1131 and issue #430 @joheredi and I are trying to use Style.camel and Style.pascal  from @azure-tools/codegen library to fix issue like method name  JWTLogin will be convert to jWTLogin, but as I commentted above there's some weird behaviour and unexpected result of Style.camel. I am still trying to find out a solution, Could you provide any suggestions? 
For issue #1134, we are trying to merge the Client and ClientContext into one file. 

@ramya-rao-a @deyaaeldeen  I will split them into two PRs seperately, as the fix for issue #1134 are pretty straight forward. But the naming issue fix probably going to take some time. I think that fix can be done after the GA, because it only impact some corner cases and we don't want this impact the GA date. let me know if that works for you ?

**Comment by ramya-rao-a** [comment]
Splitting the PR sounds like a good idea

File: test/integration/generated/nameChecker/src/operations/documents.ts [hunk 9]
```diff
@@ -143,16 +143,16 @@ export class DocumentsImpl implements Documents {
    * Autocompletes incomplete query terms based on input text and matching terms in the index.
    * @param suggesterName The name of the suggester as specified in the suggesters collection that's part
    *                      of the index definition.
-   * @param search$DONotNormalize$Text The incomplete term which should be auto-completed.
+   * @param searchDONOTNormalizeText The incomplete term which should be auto-completed.
```
Comment (qiaozha) [inline test/integration/generated/nameChecker/src/operations/documents.ts:146]: why there's a $ sign here before?

File: test/integration/generated/url2/src/operations/GET.ts [hunk 10]
```diff
@@ -6,18 +6,18 @@
  * Changes may cause incorrect behavior and will be lost if the code is regenerated.
```
Comment (qiaozha) [inline test/integration/generated/url2/src/operations/GET.ts:6]: Does this file name look okay ?

File: test/integration/generated/xmlservice/src/models/index.ts [hunk 11]
```diff
@@ -19,7 +19,7 @@ export interface RootWithRefAndNoMeta {
 /** I am a complex type with no XML node */
 export interface ComplexTypeNoMeta {
   /** The id of the res */
-  id?: string;
+  ID?: string;
```
Comment (qiaozha) [inline test/integration/generated/xmlservice/src/models/index.ts:22]: Also this one.

File: test/integration/generated/lro/src/index.ts (hunk 4 above)
Comment (deyaaeldeen) [reply test/integration/generated/lro/src/index.ts:10]: The new name feels better!

File: test/integration/generated/datalakestorage/src/models/index.ts [hunk 12]
```diff
@@ -119,7 +119,7 @@ export interface BlobPropertiesInternal {
   accessTierChangeTime?: Date;
   tagCount?: number;
   expiresOn?: Date;
-  isSealed?: boolean;
+  IsSealed?: boolean;
```
Comment (qiaozha) [inline test/integration/generated/datalakestorage/src/models/index.ts:122]: Is this 'Is' the problem ?

File: test/smoke/generated/network-resource-manager/src/operations/virtualWans.ts [hunk 13]
```diff
@@ -163,13 +163,13 @@ export class VirtualWansImpl implements VirtualWans {
    * Creates a VirtualWAN resource if it doesn't exist else updates the existing VirtualWAN.
    * @param resourceGroupName The resource group name of the VirtualWan.
    * @param virtualWANName The name of the VirtualWAN being created or updated.
-   * @param wANParameters Parameters supplied to create or update VirtualWAN.
+   * @param WANParameters Parameters supplied to create or update VirtualWAN.
```
Comment (qiaozha) [inline test/smoke/generated/network-resource-manager/src/operations/virtualWans.ts:166]: This is also a good candidate

File: src/utils/nameUtils.ts [hunk 14]
```diff
@@ -125,12 +127,17 @@ function getSuffix(nameType?: NameType) {
  */
 export function normalizeTypeName({ kind, typeName }: TypeDetails) {
   // Only Enum and Composite kinds need normalization
+  let tempTypeName = "";
+  if (typeName.endsWith("[]")) {
+    tempTypeName = "[]"
+    typeName = typeName.replace("[]", "");
+  }
```
Comment (deyaaeldeen) [inline src/utils/nameUtils.ts:134]: NIT:

`tempTypeName = "[]"` could be confusing. Could this be rewritten into a boolean variable instead that can be checked in line 136? e.g. `isArray = true;` and then 
```ts
return `${normalizeName(typeName, NameType.Interface)}${isArray ? "[]" : ""}`;
```

**Comment by sarangan12** [comment]
@qiaozha I have discussed with Jose about this PR and here is the summary:

1. Currently, the naming changes are done by 2 systems - M4 & SDK Generator. 
2. Similarly, We could override the names using 2 methods - Swagger Overrides & M4 Overrides 
3. After the discussion, we have come to the agreement that there can always be cases which might not be handled only by the M4 Logic and M4 Overrides. So, it is not safe or in our advantage to remove the naming logic we have in our SDK Generator. 
4. The edge cases where the name is different now (from what we want) are rare and could be easily handled with a combination of Swagger Overrides & M4 Overrides. 
5. So, I would recommend not to proceed with this PR and close it. The only recommended change would be to remove `$DO_NOT_NORMALIZE$` and instead could use the modeler four overrides directly (as they could be read from the host). But, again it is not a mandatory or urgent change or a change that will produce any change in the generated code.

**Comment by qiaozha** [comment]
close this PR based on @sarangan12 's comment. Thanks
